File: Code/Greedy/Add_Minimum_Number_of_Rungs_1936.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    /*
       Its just straight forward just add rung where its need and take max of rung you just added and ith rung

    */
    int addRungs(vector<int> rungs, int dist)
    {
        int ans = 0;
        int count = 0;
        for (int i = 0; i < rungs.size(); ++i)
        {
            if (rungs[i] - count > dist)
            {
                // ceil is important
                ans += ceil((rungs[i] - count - dist) * 1.0 / dist); // total rung we have to add to reach to next rung
                count += (rungs[i] - count - dist);
            }
            // we become greedy and take max because it creates a high chance to reach the next rug with optimal dist
            count = max(count, rungs[i]);
        }
        return ans;
    }
};

int main()
{
#ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif
    Solution obj;
    auto ans = obj.addRungs({1, 3, 5, 10}, 2);
    cout << ans;
    return 0;
}

File: java/Greedy_java/Add_Minimum_Number_of_Rungs_1936.java
package java.Greedy_java;

import java.util.*;
import java.io.*;

// import static java.lang.Math.max;
// import static java.lang.Math.min;
// import static java.lang.Math.abs;
// import static java.lang.System.out;

public class Add_Minimum_Number_of_Rungs_1936 {
    /*
     * Its just straight forward just add rung where its need and take max of rung
     * you just added and ith rung
     * 
     */
    public int addRungs(int[] rungs, int dist) {
        int ans = 0;
        int count = 0;
        for (int i = 0; i < rungs.length; ++i) {
            if (rungs[i] - count > dist) {
                // ceil is important
                ans += Math.ceil((rungs[i] - count - dist) * 1.0 / dist); // total rung we have to add to reach to next
                                                                          // rung
                count += (rungs[i] - count - dist);
            }
            // we become greedy and take max because it creates a high chance to reach the
            // next rug with optimal dist
            count = Math.max(count, rungs[i]);
        }
        return ans;
    }

    public static void main(String[] args) throws FileNotFoundException {
        File f = new File("output.txt");

        // Scanner ip = new Scanner(System.in);
        try (Scanner ip = new Scanner(new File("input.txt"));
                PrintWriter op = new PrintWriter(f);) {
            Add_Minimum_Number_of_Rungs_1936 obj = new Add_Minimum_Number_of_Rungs_1936();
            int nums[] = { 1, 3, 5, 10 };

            op.println(obj.addRungs(nums, 2));

        }
    }

}